### PR TITLE
Retrieve reindex history for a given content source.

### PR DESCRIPTION
**Pull request by LATaylor-guardian**

### CODE REVIEW

File: app/com/gu/floodgate/DynamoDBTable.scala [hunk 1]
```diff
@@ -62,6 +62,32 @@ trait DynamoDBTable[T] extends StrictLogging {
     promise.future
   }
 
+  def getItems(id: String, keyName: String) = {
```
Comment (cb372) [inline app/com/gu/floodgate/DynamoDBTable.scala]: Shouldn't the key name be a field of the class representing the Dynamo table?

Comment (LATaylor-guardian) [inline app/com/gu/floodgate/DynamoDBTable.scala]: Good point.

File: app/com/gu/floodgate/DynamoDBTable.scala [hunk 2]
```diff
@@ -62,6 +62,32 @@ trait DynamoDBTable[T] extends StrictLogging {
...
+
+      override def onSuccess(request: QueryRequest, result: QueryResult) = {
+        val item: List[T] = result.getItems.asScala.toList.map(f => fromItem(f.toMap))
+        promise.success(item)
+      }
+
+    }
+
+    dynamoDB.queryAsync(request, responseHandler)
+    promise.future
```
Comment (cb372) [inline app/com/gu/floodgate/DynamoDBTable.scala]: I think this "create a promise, create an async handler, make a request, return the future" process can be abstracted. Something like:

```scala
def makeAsyncDynamoRequest[Req, Resp, ReturnType](request: Req, makeReq: (req: Req, handler: AsyncHandler[Req, Resp]) => Unit, convertResult: Resp => ReturnType, errorMsg: Exception => String): Future[ReturnType] = {
  val promise = Promise[ReturnType]()
  val handler = new AsyncHandler[Req, Resp] {
    override def onError(e: Exception) = {
      logger.warn(errorMsg)
      promise.failure(e)
    }
    override def onSuccess(request: Req, result: Resp) = 
      promise.success(convertResult(result))
  }
  makeReq(request, handler)
  promise.future
}

val request = new QueryRequest().withTableName(tableName)
    .withKeyConditionExpression(s"$keyName = :$keyName")
    .withExpressionAttributeValues(Map(s":$keyName" -> new AttributeValue().withS(id)))
val makeReq = (req: QueryRequest, handler: AsyncHandler[QueryRequest, QueryResult]) => dynamoDB.queryAsync(req, handler)
val convertResult = (result: QueryResult) => result.getItems.asScala.toList.map(f => fromItem(f.toMap))
val errorMsg = (e: Exception) => s"Could not retrieve item with hash key: $id: ${e.getMessage} "
makeAsyncDynamoRequest(request, makeReq, convertResult, errorMsg)
```

(Untested, probably doesn't compile!)

File: app/com/gu/floodgate/DynamoDBTable.scala [hunk 3]
```diff
@@ -62,6 +62,32 @@ trait DynamoDBTable[T] extends StrictLogging {
     promise.future
   }
 
+  def getItems(id: String, keyName: String) = {
+    val promise = Promise[List[T]]()
+
+    val request = new QueryRequest().withTableName(tableName)
+      .withKeyConditionExpression(s"$keyName = :$keyName")
+      .withExpressionAttributeValues(
+        Map(s":$keyName" -> new AttributeValue().withS(id)))
```
Comment (cb372) [inline app/com/gu/floodgate/DynamoDBTable.scala]: Do you need to specify `withScanIndexForward(false)` to get the newest results first? I can't remember which way round the sort order works by default.

Comment (LATaylor-guardian) [inline app/com/gu/floodgate/DynamoDBTable.scala]: http://docs.aws.amazon.com/sdkfornet1/latest/apidocs/html/P_Amazon_DynamoDBv2_Model_QueryRequest_ScanIndexForward.htm

File: app/com/gu/floodgate/DynamoDBTable.scala (hunk 2 above)
Comment (LATaylor-guardian) [inline app/com/gu/floodgate/DynamoDBTable.scala]: I agree. I will do this in a separate PR.

**Pull request merged**